### PR TITLE
#45 Create discount calculation component

### PR DESCRIPTION
**Pull request by Harsha-Vardhan-24**
Hi Team,

Apologies for the delay—I got caught up with some office work.

I’ve completed the coupon component as requested. The component supports two designs, which can be toggled using the type prop (a boolean).

Additionally, I’ve implemented two separate functions that handle calculations based on the inputs provided by the user.

Here is how the design looks:

Design 1:
image

Design 2:
image

Thank you!

### CODE REVIEW

File: src/components/ui/discount.tsx [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+import { useForm, Controller } from "react-hook-form";
+import { useEffect } from "react";
+import { getCouponInfo } from "@/utils/getCouponInfo";
+
+interface DiscountFormInputs {
+  days: number;
+  discountRate: string;
+  netAmount: string;
+  sum: string;
+}
```
Comment (cryptosalomao) [inline src/components/ui/discount.tsx:21]: Any reason for only `days` being a number here?

File: src/utils/getCouponInfo.ts [hunk 2]
```diff
@@ -0,0 +1,85 @@
+import { useEffect, useState } from "react";
+
+// Custom hook to calculate coupon-related info
+export const getCouponInfo = (
+  days: number,
+  discountRate: string,
+  netAmount: string,
+  sum: string,
+  type: boolean
+) => {
```
Comment (cryptosalomao) [inline src/utils/getCouponInfo.ts:10]: Use a single object for the hooks properties

File: src/utils/getCouponInfo.ts
Comment (cryptosalomao) [inline src/utils/getCouponInfo.ts:1]: Wrong directory for a hook, and wrong file naming convention

File: src/utils/getCouponInfo.ts [hunk 3]
```diff
@@ -0,0 +1,85 @@
+import { useEffect, useState } from "react";
+
+// Custom hook to calculate coupon-related info
```
Comment (cryptosalomao) [inline src/utils/getCouponInfo.ts:3]: Unnecessary comment

File: src/utils/getCouponInfo.ts [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+  days: number,
+  discountRate: string,
+  netAmount: string,
+  sum: string,
+  type: boolean
+) => {
+  const [calculatedValue, setCalculatedValue] = useState<number | null>(null); // State to store the result
+  const [markup, setMarkup] = useState<number | null>(null); // State to store the markup value
+  const [grossAmount, setGrossAmount] = useState<number | null>(null); // State to store the gross amount
+  const [isGross, setIsGross] = useState<boolean>(true); // State to toggle between gross/net calculation
```
Comment (cryptosalomao) [inline src/utils/getCouponInfo.ts:14]: `useState<boolean>(true)` is a code smell

File: src/utils/getCouponInfo.ts [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+    const i = Number(discountRate) / 100; // Convert percentage to decimal
+    const netAmountValue = Number(netAmount.slice(1));
+
+    // Perform the calculation based on the mode (gross or net)
+    let result;
+    let calculatedMarkup: number | null = null; // Store markup here
+
+    if (isGross) {
+      result = timeDiff * Math.exp(i) * netAmountValue; // Gross amount formula
+      setGrossAmount(result); // Set gross amount
```
Comment (cryptosalomao) [inline src/utils/getCouponInfo.ts:41]: Unnecessary comment

File: src/utils/getCouponInfo.ts [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+      calculatedMarkup = result - netAmountValue; // Markup = gross amount - net amount
```
Comment (cryptosalomao) [inline src/utils/getCouponInfo.ts:42]: Unnecessary comment

File: src/utils/getCouponInfo.ts [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+    } else {
+      result = netAmountValue / (timeDiff * Math.exp(i)); // Net amount formula
+      setGrossAmount(null); // Set gross amount to null for net calculation
+      calculatedMarkup = 0; // No markup for net calculation
+    }
+
+    setCalculatedValue(result); // Set calculated value (gross or net)
+    setMarkup(calculatedMarkup); // Set calculated markup
```
Comment (cryptosalomao) [inline src/utils/getCouponInfo.ts:50]: Unnecessary comments

File: src/utils/getDates.ts [hunk 8]
```diff
@@ -0,0 +1,20 @@
+export const getDates = (dateRange: number) => {
+  const today = new Date();
+  const futureDate = new Date();
+
+  // Ensure dateRange is a number
+  const numericDateRange = Number(dateRange);
```
Comment (cryptosalomao) [inline src/utils/getDates.ts:6]: `dateRange` parameter is already type checked against numeric values

File: src/utils/getDates.ts [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+
+  // Add the range in days
```
Comment (cryptosalomao) [inline src/utils/getDates.ts:8]: Unnecessary comment

File: src/utils/getDates.ts [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+  futureDate.setDate(today.getDate() + numericDateRange);
+
+  const formatDate = (date: Date) =>
+    `${date.getDate().toString().padStart(2, "0")}-${(date.getMonth() + 1)
+      .toString()
+      .padStart(2, "0")}-${date.getFullYear()}`;
```
Comment (cryptosalomao) [inline src/utils/getDates.ts:14]: There are cleaner ways to achieve the desired format. Please check `Intl.DateTimeFormat` doc below:

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/DateTimeFormat

File: src/utils/getCouponInfo.ts [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+    const sumValue = Number(sum.slice(1));
+
+    // Adjust discount based on days (e.g., prorate the discount if it's based on time)
+    // Example: Apply a different discount factor based on the number of days
+    const timeFactor = days / 365; // For simplicity, adjust by days to years. You can adjust this logic as needed.
+    const discount = sumValue * discountRateValue * timeFactor; // Adjusted discount
+    const discountedValue = sumValue - discount; // Final discounted value after applying discount
+
+    setCalculatedValue(discountedValue); // Set the discounted value
+    setMarkup(discount); // Set the discount as markup
```
Comment (cryptosalomao) [inline src/utils/getCouponInfo.ts:71]: Unnecessary comments

File: src/components/ui/discount.tsx [hunk 12]
```diff
@@ -0,0 +1,269 @@
...
+
+export const DiscountCoupon = ({ type }: { type: boolean }) => {
```
Comment (cryptosalomao) [inline src/components/ui/discount.tsx:23]: What exactly does `type` do here?

File: src/components/ui/discount.tsx [hunk 13]
```diff
@@ -0,0 +1,269 @@
...
+      netAmount: "",
+    },
+  });
+  const { days, discountRate, netAmount, sum } = watch();
+  const { formattedToday, formattedFutureDate } = getDates(days);
+  const { calculatedValue, setIsGross, isGross, markup, grossAmount } =
+    getCouponInfo(days, discountRate, netAmount, sum, type);
+  const onSubmit = (data: DiscountFormInputs) => {
+    console.log(data);
+  };
```
Comment (cryptosalomao) [inline src/components/ui/discount.tsx:42]: There should be a way to trigger a callback for the calculation submission here

**Comment by mtbitcr** [comment]
Hi @Harsha-Vardhan-24. I can see that this is probably code generated using ChatGPT. In the future, please don't use AI to write code. 
Next time PRs of this quality will be rejected immediately. This time, redo it the way @cryptosalomao wrote it for you.

**Comment by mjkeaton** [comment]
> [...] Next time PRs of this quality will be rejected immediately. This time, redo it the way @cryptosalomao wrote it for you.

I am going to close this PR for now. Please feel free to reopen anytime you think it is of acceptable quality. :pray:

**Pull request closed without merging**